website/addons/evernote/model.py

```diff
@@ -0,0 +1,128 @@
+    """
+    First cut at the Evernote provider
+
+    """
+    name = 'Evernote'
+    short_name = 'evernote'
+
+    # Evernote has a sandbox and a production service
+    # with different endpoints
+
+    client_id = settings.EVERNOTE_CLIENT_ID
+    client_secret = settings.EVERNOTE_CLIENT_SECRET
+    _sandbox = settings.EVERNOTE_SANDBOX
+
+    BASE_URL = "https://www.evernote.com" if not _sandbox  \
+           else "https://sandbox.evernote.com"
+
+    auth_url_base = '{}/OAuth.action'.format(BASE_URL)
+    request_token_url = '{}/oauth'.format(BASE_URL)
+    callback_url = '{}/oauth'.format(BASE_URL)
+
+    _oauth_version = OAUTH1
+
+    def handle_callback(self, response):
+        """View called when the Oauth flow is completed. Adds a new BoxUserSettings
+        record to the user and saves the user's access token and account info.
+        """
+
+        client = utils.get_evernote_client(token=response.get('oauth_token'))
+
+        userStore = client.get_user_store()
+        user = userStore.getUser()
+
+        return {
+            'provider_id': user.id,  # or user.username
+            'display_name': user.name,
+            'profile_url': ''
+        }
+
+class EvernoteUserSettings(AddonOAuthUserSettingsBase):
+    oauth_provider = Evernote
+    serializer = EvernoteSerializer
+
+
+class EvernoteNodeSettings(StorageAddonBase, AddonOAuthNodeSettingsBase):
+    oauth_provider = Evernote
+    serializer = EvernoteSerializer
+
+    folder_id = fields.StringField(default=None)
+    folder_name = fields.StringField()
+    folder_path = fields.StringField()
+
+    # to hold data about current Notebook
+    _folder_data = None
+
+    def set_user_auth(self, user_settings):
+        """Import a user's Evernote authentication and create a NodeLog.
+        :param EvernoteUserSettings user_settings: The user settings to link.
+        """
+        self.user_settings = user_settings
+        #nodelogger = BoxNodeLogger(node=self.owner, auth=Auth(user_settings.owner))
+        #nodelogger.log(action="node_authorized", save=True)
+
+    def set_folder(self, folder_id, auth):
+        self.folder_id = str(folder_id)
+        self._update_folder_data()
+        self.save()
+
+        if not self.complete:
+            self.user_settings.grant_oauth_access(
+                node=self.owner,
+                external_account=self.external_account,
+                metadata={'folder': self.folder_id}
+            )
+            self.user_settings.save()
+
+        # Add log to node
+        #nodelogger = BoxNodeLogger(node=self.owner, auth=auth)
+        #nodelogger.log(action="folder_selected", save=True)
+
+    def _update_folder_data(self):
+        """
+        given the folder_id, compute folder_name, folder_path
+        """
+        if self.folder_id is None:
+            return None
+
+        if not self._folder_data:
+            try:
+                # don't know whether we'll need to do something like this
```

Comment:
According to the [EverNote docs](https://dev.evernote.com/doc/articles/authentication.php), it looks like you won't be able to:

>Each access token has an expiration date. By default, the duration of access token validity is 1 year from the date of issue. The user can alter this duration to 1 day, 1 week or 1 month. In all these cases (including tokens valid for 1 year), the expiration date will be included as the parameter edam_expires. The value of this parameter will be a standard UNIX timestamp in units of milliseconds indicating the date and time of expiration.

The don't return a refresh token for API applications, and [`UserStore.refreshAuthentication` is only available for internal applications](https://dev.evernote.com/doc/reference/UserStore.html#Fn_UserStore_refreshAuthentication)